Kitodo-FileManagement/src/main/java/org/kitodo/filemanagement/locking/LockManagement.java

```diff
@@ -0,0 +1,511 @@
+     */
+    final ImmutableReadFileManagement immutableReadFileManagement = new ImmutableReadFileManagement();
+
+    /**
+     * Variable for accessing the stream management.
+     */
+    private final StreamManagement streamManagement = new StreamManagement();
+
+    /**
+     * Checks for a group of locks if they can be returned.
+     *
+     * @param uris
+     *            the URIs on the locks to be returned
+     * @throws IllegalStateException
+     *             if there is still a stream flowing over the lock
+     */
+    private void canClose(Collection<URI> uris) {
+        String exceptionMessagePrefix = "The lock cannot be released because there are still streams flowing for the following URIs: ";
+        String exceptionMessage = uris.parallelStream().filter(uri -> streamManagement.isKnowingAnOpenStreamTo(uri))
+                .map(uri -> uri.toString()).collect(Collectors.joining(", ", exceptionMessagePrefix, ""));
+        if (exceptionMessage.length() > exceptionMessagePrefix.length()) {
+            throw new IllegalStateException(exceptionMessage);
+        }
+    }
+
+    /**
+     * Checks the ability to lock in a parallel fashion. The called test method
+     * returns {@code null} if successful, and this method removes the
+     * {@code null}s from the stream, leaving a stream of conflicts at the end.
+     * If all locks could be granted, the stream will no longer contain any
+     * elements here, indicating that all requested locks can be granted.
+     *
+     * @param user
+     *            the user requesting the locks
+     * @param lockRequests
+     *            a stream of desired locks
+     * @return a stream of locks that can not be granted as pairs of URI and the
+     *         names of those users who hold locks that prevent the requested
+     *         lock from being granted
+     */
+    private Stream<Pair<URI, Collection<String>>> checkLockability(String user,
+            Stream<Entry<URI, LockingMode>> lockRequests) {
+        return lockRequests.map(entry -> checkLockability(user, entry.getKey(), entry.getValue()))
+                .filter(Objects::nonNull);
+    }
+
+    /**
+     * Checks if a user can get a lock on an URI. This method returns
+     * {@code null} if successful, and the calling method removes the zeroes
+     * from the stream, leaving a stream of conflicts at the end.
+     *
+     * @param user
+     *            user who wants the lock
+     * @param uri
+     *            URI for which the lock is desired
+     * @param requestedLock
+     *            which lock is desired
+     * @return {@code null} on success, otherwise a pair of the URI and all
+     *         users who have locks that conflict with the requested lock
+     */
+    private Pair<URI, Collection<String>> checkLockability(String user, URI uri, LockingMode requestedLock) {
+        if (grantedPermissions.get(uri) == null) {
+            return null;
```

Comment:
Null return should be avoided. Pair also should be avoided. For such a cases classes are better. In that case you could return class with empty collection of strings.